EKF/mag_fusion.cpp

```diff
@@ -435,304 +435,212 @@ void Ekf::fuseMag()
-			// Values from yaw_input_312.c file produced by https://github.com/PX4/ecl/blob/master/matlab/scripts/Inertial%20Nav%20EKF/quat2yaw312.m
-			float Tbn_0_1_neg = 2.0f*(_ev_sample_delayed.quat(0)*_ev_sample_delayed.quat(3)-_ev_sample_delayed.quat(1)*_ev_sample_delayed.quat(2));
-			float Tbn_1_1 = sq(_ev_sample_delayed.quat(0))-sq(_ev_sample_delayed.quat(1))+sq(_ev_sample_delayed.quat(2))-sq(_ev_sample_delayed.quat(3));
-			measured_hdg = atan2f(Tbn_0_1_neg,Tbn_1_1);
-
-		} else if (_mag_use_inhibit) {
-			// Special case where we either use the current or last known stationary value
-			// so set to current value as a safe default
-			measured_hdg = predicted_hdg;
-
-		} else {
-			// Should not be doing yaw fusion
-			return;
+		return;
 
-		}
 	}
 
-	// Calculate the observation variance
-	if (_control_status.flags.mag_hdg) {
-		// using magnetic heading tuning parameter
-		R_YAW = sq(fmaxf(_params.mag_heading_noise, 1.0e-2f));
-
-	} else if (_control_status.flags.ev_yaw) {
-		// using error estimate from external vision data
-		R_YAW = fmaxf(_ev_sample_delayed.angVar, sq(1.0e-2f));
-
+	if (zero_innovation) {
+		_heading_innov = 0.0f;
 	} else {
-		// default value
-		R_YAW = 0.01f;
+		// calculate the the innovation and wrap to the interval between +-pi
+		_heading_innov = wrap_pi(atan2f(-_R_to_earth(0, 1), _R_to_earth(1, 1)) - measurement);
 	}
 
-	// wrap the heading to the interval between +-pi
-	measured_hdg = wrap_pi(measured_hdg);
-
-	// calculate the innovation and define the innovation gate
+	// define the innovation gate size
 	float innov_gate = math::max(_params.heading_innov_gate, 1.0f);
-	if (_mag_use_inhibit) {
-		// The magnetometer cannot be trusted but we need to fuse a heading to prevent a badly
-		// conditioned covariance matrix developing over time.
-		if (!_control_status.flags.vehicle_at_rest) {
-			// Vehicle is not at rest so fuse a zero innovation and record the
-			// predicted heading to use as an observation when movement ceases.
-			_heading_innov = 0.0f;
-			_last_static_yaw = predicted_hdg;
-
-		} else {
-			// Vehicle is at rest so use the last moving prediction as an observation
-			// to prevent the heading from drifting and to enable yaw gyro bias learning
-			// before takeoff.
-			_heading_innov = predicted_hdg - _last_static_yaw;
-			innov_gate = 5.0f;
 
-		}
-	} else {
-		_heading_innov = predicted_hdg - measured_hdg;
-		_last_static_yaw = predicted_hdg;
-
-	}
-	_mag_use_inhibit_prev = _mag_use_inhibit;
-
-	// wrap the innovation to the interval between +-pi
-	_heading_innov = wrap_pi(_heading_innov);
+	// Update the quaternion states and covariance matrix
+	updateQuaternion(_heading_innov, R_YAW, innov_gate, H_YAW);
+}
 
+// update quaternion states and covariances using the yaw innovation, yaw observation variance and yaw Jacobian
+void Ekf::updateQuaternion(const float innovation, const float variance, const float gate_sigma, const float *yaw_jacobian)
```

Comment:
```suggestion
void Ekf::updateQuaternion(const float innovation, const float variance, const float gate_sigma, const float (&yaw_jacobian)[4])
```
Passing by reference is safer.